File: python/tests/test_checksum_sweep.py
"""Checksum single-substitution detection sweep, spec
IMPLEMENTATION_TEST_SUITE.md section 6.3.

For each checksummed frozen tier (Light, Medium and Heavy): over a sample of
random bodies, substitute every body position with every other canonical
symbol and record whether the checksum still matches. All three tiers ship
two checksum symbols and must show total detection: zero misses.

Two run levels share one implementation, matching tests/test_soak.py:

- CI subset (default): 2,000 sampled bodies per tier. Runs inside the normal
  pytest run.
- Full sweep (opt-in): 100,000 sampled bodies per tier, the spec's release
  gate. Selected with BASEH_SOAK=1; skipped cleanly otherwise.

Environment overrides (all optional):

- BASEH_SOAK=1          run the full 100,000-body sweep
- BASEH_SOAK_CHECKSUM   override the sampled body count
- BASEH_SOAK_SEED       override the sample seed (default 42)
"""

import os
import random
import unittest

from baseh import baseh_heavy_v1, baseh_light_v1, baseh_medium_v1
from baseh.basen import alphabet_index
from baseh.checksum import checksum_value
from baseh.profile import prepare_profile

_SOAK = os.environ.get("BASEH_SOAK") == "1"
_SEED = int(os.environ.get("BASEH_SOAK_SEED", "42"))
_BODY_COUNT = int(os.environ.get("BASEH_SOAK_CHECKSUM", "100000" if _SOAK else "2000"))

_TIERS = (
    ("baseh-light-v1", baseh_light_v1),
    ("baseh-medium-v1", baseh_medium_v1),
    ("baseh-heavy-v1", baseh_heavy_v1),
)


class _SweepBase(unittest.TestCase):
    def _sweep(self, name, build, body_count):
        profile = prepare_profile(build())
        alphabet = profile.body_alphabet_norm
        index = alphabet_index(alphabet)
        k = profile.checksum_length
        body_length = profile.body_length
        self.assertEqual(k, 2, f"{name} is expected to ship two checksum symbols")
        rng = random.Random(f"{_SEED}:{name}")
        misses = []
        checked = 0
        for _ in range(body_count):
            body = "".join(rng.choice(alphabet) for _ in range(body_length))
            before = checksum_value(profile, body, index, k)
            for pos in range(body_length):
                current = index[body[pos]]
                for value in range(len(alphabet)):
                    if value == current:
                        continue
                    candidate = body[:pos] + alphabet[value] + body[pos + 1 :]
                    checked += 1
                    if checksum_value(profile, candidate, index, k) == before:
                        misses.append((body, pos, alphabet[value]))
        self.assertEqual(
            misses,
            [],
            f"{name}: {len(misses)} undetected substitutions in "
            f"{checked} checks (first: {misses[:1]})",
        )

    def _run(self, body_count):
        print(f"checksum sweep: seed={_SEED} bodies={body_count} per tier")
        for name, build in _TIERS:
            with self.subTest(profile=name):
                self._sweep(name, build, body_count)


class TestChecksumSweepCISubset(_SweepBase):
    """Default CI level: 2,000 sampled bodies per checksummed frozen tier."""

    def test_substitution_subset(self):
        self._run(min(_BODY_COUNT, 2000))


@unittest.skipUnless(_SOAK, "full checksum sweep is opt-in: set BASEH_SOAK=1")
class TestChecksumSweepFull(_SweepBase):
    """Full 100,000-body release-gate sweep, spec section 6.3."""

    def test_substitution_full(self):
        self._run(_BODY_COUNT)


if __name__ == "__main__":
    unittest.main()

File: ruby/test/test_checksum_sweep.rb
# frozen_string_literal: true

require "minitest/autorun"
require "baseh"

# Single-substitution checksum detection sweep, spec/IMPLEMENTATION_TEST_SUITE.md
# section 6. For each checksummed frozen tier (Light, Medium and Heavy) every
# single-symbol substitution of a sampled body must change the checksum:
# zero misses, per spec 6.3.
#
# Two run levels, same idiom as test_soak.rb:
#   * CI subset (default test run): CI_BODY_COUNT sampled bodies per tier.
#   * Full sweep: BASEH_SOAK=1 runs the spec's 100,000 sampled bodies per
#     tier (overridable via BASEH_SOAK_BODIES). Without BASEH_SOAK the
#     soak-level test skips cleanly.
class TestChecksumSweep < Minitest::Test
  SEED = 42

  CI_BODY_COUNT = 500
  SOAK_BODY_COUNT = 100_000

  def soak?
    ENV["BASEH_SOAK"] == "1"
  end

  def soak_body_count
    (ENV["BASEH_SOAK_BODIES"] || SOAK_BODY_COUNT).to_i
  end

  def checksummed_tiers
    [
      ["baseh-light-v1", Baseh.baseh_light_v1],
      ["baseh-medium-v1", Baseh.baseh_medium_v1],
      ["baseh-heavy-v1", Baseh.baseh_heavy_v1]
    ]
  end

  # Sweeps `count` seeded random bodies; returns the miss count (substitutions
  # whose checksum did not change).
  def sweep(profile, count)
    codec = Baseh::Baseh.new(profile)
    prepared = codec.profile
    index = Baseh::BaseN.alphabet_index(prepared.body_alphabet)
    alphabet = prepared.body_alphabet.chars
    rng = Random.new(SEED)
    misses = 0
    count.times do
      body = Array.new(prepared.body_length) { alphabet[rng.rand(alphabet.size)] }.join
      expected = Baseh::Checksum.calculate_checksum(prepared, body, index)
      chars = body.chars
      chars.each_index do |pos|
        alphabet.each do |replacement|
          next if replacement == chars[pos]

          candidate = chars.dup
          candidate[pos] = replacement
          substituted = Baseh::Checksum.calculate_checksum(prepared, candidate.join, index)
          misses += 1 if substituted == expected
        end
      end
    end
    misses
  end

  def run_sweep(count)
    checksummed_tiers.each do |name, profile|
      misses = sweep(profile, count)
      assert_equal 0, misses,
                   "#{name}: #{misses} single substitutions went undetected " \
                   "over #{count} sampled bodies (spec 6.3 requires zero)"
    end
  end

  # CI subset: always runs, must finish in seconds.
  def test_single_substitution_sweep_ci_subset
    run_sweep(CI_BODY_COUNT)
  end

  # Full 100,000-body sweep of test-suite section 6: opt-in via BASEH_SOAK=1.
  def test_single_substitution_sweep_soak
    skip "full checksum sweep: set BASEH_SOAK=1 to run" unless soak?

    run_sweep(soak_body_count)
  end
end
